core/src/spclient.rs

```diff
@@ -1 +1,308 @@
+                }
+            }
+
+            // Reconnection logic: drop the current access point if we are experiencing issues.
+            // This will cause the next call to base_url() to resolve a new one.
+            if let Err(SpClientError::Network(ref network_error)) = last_response {
+                match network_error {
+                    HttpClientError::Response(_) | HttpClientError::Request(_) => {
+                        // Keep trying the current access point three times before dropping it.
+                        if tries % 3 == 0 {
+                            self.flush_accesspoint().await
+                        }
+                    }
+                    _ => break, // if we can't build the request now, then we won't ever
+                }
+            }
+
+            // When retrying, avoid hammering the Spotify infrastructure by sleeping a while.
+            // The backoff time is chosen randomly from an ever-increasing range.
+            let max_seconds = u64::pow(tries as u64, 2) * 3;
+            let backoff = Duration::from_secs(rand::thread_rng().gen_range(1..=max_seconds));
+            warn!(
+                "Unable to complete API request, waiting {} seconds before retrying...",
+                backoff.as_secs(),
+            );
+            debug!("Error was: {:?}", last_response);
+            tokio::time::sleep(backoff).await;
+        }
+
+        last_response
+    }
+
+    pub async fn put_connect_state(
+        &self,
+        connection_id: String,
+        state: PutStateRequest,
+    ) -> SpClientResult {
+        let endpoint = format!("/connect-state/v1/devices/{}", self.session().device_id());
+
+        let mut headers = HeaderMap::new();
+        headers.insert("X-Spotify-Connection-Id", connection_id.parse()?);
+
+        self.request_with_protobuf(&Method::PUT, &endpoint, Some(headers), &state)
+            .await
+    }
+
+    pub async fn get_metadata(&self, scope: &str, id: SpotifyId) -> SpClientResult {
+        let endpoint = format!("/metadata/4/{}/{}", scope, id.to_base16());
+        self.request(&Method::GET, &endpoint, None, None).await
+    }
+
+    pub async fn get_track_metadata(&self, track_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("track", track_id).await
+    }
+
+    pub async fn get_episode_metadata(&self, episode_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("episode", episode_id).await
+    }
+
+    pub async fn get_album_metadata(&self, album_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("album", album_id).await
+    }
+
+    pub async fn get_artist_metadata(&self, artist_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("artist", artist_id).await
+    }
+
+    pub async fn get_show_metadata(&self, show_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("show", show_id).await
+    }
+
+    pub async fn get_lyrics(&self, track_id: SpotifyId) -> SpClientResult {
```

Comment:
Should we return a parsed version of the json instead?

Here are structs doing the serde deserialization:

```rust
#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
struct LyricsReply {
    colors: Colors,
    has_vocal_removal: bool,
    lyrics: Lyrics,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
struct Colors {
    active_color: i32,
    bright_color: i32,
    dark_color: i32,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct Lyrics {
    kind: String,
    language: String,
    lines: Vec<Line>,
    provider: String,
    track_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
struct Line {
    time: u32,
    words: Vec<Word>,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
struct Word {
    string: String,
}
```

Reply:
Though I'm thankful for posting these structs, it's a design decision to return `Bytes` so that the caller can do with it what they want, with a JSON parser of their choosing (or none at all). `spclient` is pretty low-level that way.

As I type this I am pondering if this should be part of `librespot-metadata` or not.

Reply:
Sure no problem, feel free to ignore or re-use as you see fit.